### PR TITLE
Extend Polonius fact generation for (some) move tracking

### PR DESCRIPTION
**Pull request by amandasystems**
This PR will extend rustc to emit facts used for tracking moves and initialization in Polonius. It is most likely the final part of my master's thesis work.

### CODE REVIEW

**Comment by albins** [comment]
?r @nikomatsakis 

(for now!)

**Comment by albins** [comment]
r? @nikomatsakis 

(for now!)

**Comment by joelpalmer** [comment]
Ping from triage. This PR is in need of review. Should it be re-assigned? @nikomatsakis @cramertj 
cc: @albins

**Comment by albins** [comment]
> Ping from triage. This PR is in need of review. Should it be re-assigned? @nikomatsakis @cramertj
> cc: @albins

@joelpalmer Thanks for asking, but it's still very WIP; the fact generation is currently slightly wrong. I expect to fix it this week, but there are probably more issues than that hiding behind that issue.

However, before this PR can be merged, Polonius needs to be updated from another branch (I will rebase and fixup the commit using my local Polonius to in stead pull the new release).

So in short; I suspect @nikomatsakis might be back from their vacation before I'm done. If not, I'll ask around in the Polonius working group. Please don't reassign for now, thank you!

**Comment by albins** [comment]
This pull request is waiting for https://github.com/rust-lang/polonius/pull/110 to be merged into Polonius.

**Comment by hdhoang** [comment]
Thank you for your update, `triage-wg` have updated this PR's label accordingly.

**Comment by albins** [comment]
r? @nikomatsakis

Ready for review, as https://github.com/rust-lang/polonius/pull/110 has landed! 🎉

This should be enough facts to implement move analysis *and* liveness tracking in Polonius, meaning that an entire new part of the Polonius analysis is finally fully unlocked!

File: src/librustc_mir/borrow_check/nll/mod.rs [hunk 1]
```diff
@@ -69,6 +69,61 @@ pub(in crate::borrow_check) fn replace_regions_in_mir<'cx, 'tcx>(
...
 }
 
+
+// This function populates an AllFacts instance with base facts related to
+// MovePaths and needed for the move analysis.
+fn populate_polonius_move_facts(all_facts: &mut AllFacts, move_data: &MoveData<'_>, location_table: &LocationTable, body: &Body<'_>) {
+    all_facts.var_starts_path.extend(move_data.rev_lookup.iter_locals_enumerated().map(|(v, &m)| (v, m)));
+
+    for (idx, move_path) in move_data.move_paths.iter_enumerated() {
+        all_facts.parent.extend(move_path.parents(&move_data.move_paths).iter().map(|&parent| (parent, idx)));
```
Comment (nikomatsakis) [inline src/librustc_mir/borrow_check/nll/mod.rs]: as noted in the polonius PR, ultimately I think we should leave the transitive computation here to polonius. But I guess I don't wnat to do that in this PR.

File: src/librustc_mir/borrow_check/nll/mod.rs [hunk 2]
```diff
@@ -69,6 +69,61 @@ pub(in crate::borrow_check) fn replace_regions_in_mir<'cx, 'tcx>(
...
+
+                    for &successor in block_data.terminator().successors() {
+                        if body[successor].is_cleanup {
+                            continue;
+                        }
+
+                        // The initialization happened in (or rather, when arriving at)
+                        // the successors, but not in the unwind block.
+                        let first_statement = Location { block: successor, statement_index: 0};
+                        all_facts.initialized_at.push((init.path, location_table.start_index(first_statement)));
```
Comment (nikomatsakis) [inline src/librustc_mir/borrow_check/nll/mod.rs]: Interesting quirk. I guess this makes sense.

File: src/librustc_mir/borrow_check/nll/type_check/liveness/polonius.rs [hunk 3]
```diff
@@ -36,9 +41,21 @@ impl LivenessPointFactsExtractor<'_> {
         debug!("LivenessFactsExtractor::insert_drop_use()");
         self.var_drop_used.push((local, self.location_to_index(location)));
     }
+
+    fn insert_path_access(&mut self, path: MovePathIndex, location: Location) {
+        debug!("LivenessFactsExtractor::insert_path_access({:?}, {:?})", path, location);
+        self.path_accessed_at.push((path, self.location_to_index(location)));
+    }
+
+    fn place_to_mpi(&self, place: &Place<'_>) -> Option<MovePathIndex> {
```
Comment (nikomatsakis) [inline src/librustc_mir/borrow_check/nll/type_check/liveness/polonius.rs]: might be nice to have some docs here -- when does this return `None`?

Comment (nikomatsakis) [reply src/librustc_mir/borrow_check/nll/type_check/liveness/polonius.rs]: I guess it is when the `place` is rooted in a static, though

**Review by nikomatsakis — CHANGES_REQUESTED** [review]

**Comment by nikomatsakis** [comment]
I'm not going to block on those nits. Doesn't seem worth it.

File: src/librustc_mir/borrow_check/nll/mod.rs (hunk 2 above)
Comment (albins) [reply src/librustc_mir/borrow_check/nll/mod.rs]: Let's just say I scratched my head for quite some time before I figured out why this happened.

**Pull request merged**